File: 331-verify-preorder-serialization-of-a-binary-tree/verify-preorder-serialization-of-a-binary-tree.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        nodes = preorder.split(',')
        stack = []

        for node in nodes:
            stack.append(node)

            while (
                len(stack) >= 3
                and stack[-1] == '#'
                and stack[-2] == '#'
                and stack[-3] != '#'
            ):
                stack.pop()
                stack.pop()
                stack.pop()

                stack.append('#')

        return stack == ['#']

File: 331-verify-preorder-serialization-of-a-binary-tree/verify-preorder-serialization-of-a-binary-tree.java
import java.util.*;

class Solution {
    public boolean isValidSerialization(String preorder) {

        String[] nodes = preorder.split(",");
        Stack<String> stack = new Stack<>();

        for (String node : nodes) {

            stack.push(node);

            while (
                stack.size() >= 3 &&
                stack.get(stack.size() - 1).equals("#") &&
                stack.get(stack.size() - 2).equals("#") &&
                !stack.get(stack.size() - 3).equals("#")
            ) {
                stack.pop();
                stack.pop();
                stack.pop();

                stack.push("#");
            }
        }

        return stack.size() == 1 && stack.peek().equals("#");
    }
}